utils/oscap.8
.TH OSCAP "8" "March 2021" "Red Hat" "System Administration Utilities"

.SH NAME
oscap \- OpenSCAP command line tool

.SH SYNOPSIS
\fBoscap\fR [\fIgeneral-options\fR] \fBmodule\fR operation [\fIoperation-options-and-arguments\fR]

.SH DESCRIPTION
\fBoscap\fP is Security Content Automation Protocol (SCAP) toolkit based on OpenSCAP library. It provides various functions for different SCAP specifications (modules).

OpenSCAP tool claims to provide capabilities of Authenticated Configuration Scanner and Authenticated Vulnerability Scanner as defined by The National Institute of Standards and Technology.

.SH GENERAL OPTIONS
.TP
\fB\-V, -\-version\fR
Print supported SCAP specifications, location of schema files, schematron files, CPE files, probes and supported OVAL objects.
Displays a list of inbuilt CPE names.
.TP
\fB\-h, \-\-help\fR
Help screen.

.SH MODULES
.TP
\fBinfo\fR
Determine type and print information about a file.
.TP
\fBxccdf\fR
The eXtensible Configuration Checklist Description Format.
.TP
\fBoval\fR
Open Vulnerability and Assessment Language.
.TP
\fBds\fR
SCAP Data Stream
.TP
\fBcpe\fR
Common Platform Enumeration.

.SH COMMON OPTIONS FOR ALL MODULES
.RE
.TP
\fB\-\-verbose VERBOSITY_LEVEL\fR
.RS
Turn on verbose mode at specified verbosity level. VERBOSITY_LEVEL is one of: DEVEL, INFO, WARNING, ERROR.
.RE
.TP
\fB\-\-verbose-log-file FILE\fR
.RS
Set filename to write additional information.
.RE

.SH INFO OPERATIONS
.TP
[\fIoptions\fR] any-scap-file.xml
.RS
This module prints information about SCAP content in a file specified on a command line. It determines SCAP content type, specification version, date of creation, date of import and so on. Info module doesn't require any additional operation switch.

For XCCDF or SCAP source data stream files, the info module prints out IDs of incorporated profiles, components, and data streams. These IDs can be used to specify the target for evaluation. Use options --profile, --xccdf-id (or --oval-id), and --datastream-id respectively.
.PP
\fB\-\-fetch-remote-resources\fR
.RS
Allow download of remote components referenced from data stream.
.RE
.TP
\fB\-\-list-rules\fR
.RS
Print IDs of XCCDF rules that are selected by the given profile, one per line. This option must be used together with \fB\-\-profile\fR. The output is machine-readable and contains only rule IDs with no additional decoration.
.RE
.TP
\fB\-\-list-vars\fR
.RS
Print XCCDF Value IDs and their resolved values for the given profile, one per line, tab-separated. This option must be used together with \fB\-\-profile\fR. The output is machine-readable.
.RE
.TP
\fB\-\-local-files DIRECTORY\fR
.RS
Instead of downloading remote data stream components from the network, use data stream components stored locally as files in the given directory. In place of the remote data stream component OpenSCAP will attempt to use a file whose file name is equal to @name attribute of the uri element within the catalog element within the component-ref element in the data stream if such file exists.
This option works only with SCAP 1.3 source data streams. It can't be used with SCAP 1.2 source data streams.
.RE
.TP
\fB\-\-profile PROFILE\fR
.RS
Show info of the profile with the given ID.
.RE
.TP
\fB\-\-profiles\fR
.RS
Show profiles from the input file in the <id>:<title> format, one line per profile.
.RE
.TP
\fB\-\-references\fR
.RS
Show list of references available in each benchmark.
.RE

.SH XCCDF OPERATIONS
.TP
.B \fBeval\fR [\fIoptions\fR] INPUT_FILE [\fIoval-definitions-files\fR]
.RS
Perform evaluation of XCCDF document file given as INPUT_FILE. Print result of each rule to standard output, including rule title, rule id and security identifier(CCE). Optionally you can give an SCAP source data stream as the INPUT_FILE instead of an XCCDF file (see --datastream-id).
.PP
oscap returns 0 if all rules pass. If there is an error during evaluation, the return code is 1. If there is at least one rule with either fail or unknown result, oscap finishes with return code 2.
.PP
.PP
Unless --skip-validation is used, the INPUT_FILE is validated using XSD schemas (depending on document type of INPUT_FILE) and rejected if invalid.
.PP
You may specify OVAL Definition files as the last parameter, XCCDF evaluation will then proceed only with those specified files. Otherwise, when \fIoval-definitions-files\fR parameter is missing, oscap tool will try to load all OVAL Definition files referenced from XCCDF automatically (search in the same path as XCCDF).
.PP
.TP
\fB\-\-profile PROFILE\fR
.RS
Select a particular profile from XCCDF document. If "(all)" is given a virtual profile that selects all groups and rules will be used.
.RE
.TP
\fB\-\-rule RULE\fR
.RS
Select a particular rule from XCCDF document. Only this rule will be evaluated. Any other rules required by this rule won't be evaluated. Rule will use values according to the selected profile. If no profile is selected, default values are used. This option can be used multiple times to specify multiple rules at once.
.RE
.TP
\fB\-\-skip-rule RULE\fR
.RS
Skip a particular rule from XCCDF document. This option can be used multiple times to skip multiple rules at once.
.RE
.TP
\fB\-\-reference NAME:IDENTIFIER\fR
.RS
Evaluate rules matching the given reference. This option can be used to check for evaluating compliance with a particular policy requirement. Rules that match the reference will be evaluated, other rules will be skipped and a result of notselected will be returned. The argument needs to be provided in the form NAME:IDENTIFIER (colon is a separator), where NAME is the reference name used as a text of xccdf:reference elements that are child of xccdf:Benchmark element and IDENTIFIER is a text of xccdf:reference element that is child of a particular xccdf:Rule element which has the same @href value as the xccdf:reference element of the NAME value. This feature requires SCAP content to be formed in the aforementioned way. For example: --reference "cis:4.1.3.19", --reference "anssi:BP28(R29)"
List of references available in the given benchmark can be obtained by running the oscap info --references command.
.RE
.TP
\fB\-\-tailoring-file TAILORING_FILE\fR
.RS
Use given file for XCCDF tailoring. Select profile from tailoring file to apply using --profile. If both --tailoring-file and --tailoring-id are specified, --tailoring-file takes priority.
.RE
.TP
\fB\-\-tailoring-id COMPONENT_REF_ID\fR
.RS
Use tailoring component in input source data stream for XCCDF tailoring. The tailoring component must be specified by its Ref-ID (value of component-ref/@id attribute in input source data stream). Select profile from tailoring component to apply using --profile. If both --tailoring-file and --tailoring-id are specified, --tailoring-file takes priority.
.RE
.TP
\fB\-\-cpe CPE_FILE\fR
.RS
Use given CPE dictionary or language (auto-detected) for applicability checks. (Some CPE names are provided by openscap, see oscap --version for inbuilt CPE names)
.RE
.TP
\fB\-\-results FILE\fR
.RS
Write XCCDF results into FILE.
.RE
.TP
\fB\-\-results-arf FILE\fR
.RS
Writes results to a given FILE in Asset Reporting Format. It is recommended to use this option instead of --results when dealing with data streams.
.RE
.TP
\fB\-\-stig-viewer FILE\fR
.RS
Writes XCCDF results into FILE. The rule result IDs in FILE are modified according to STIG references in evaluated content. The FILE can be simply imported into DISA STIG Viewer. See \fIhttps://public.cyber.mil/stigs/srg-stig-tools/\fR for information about DISA STIG Viewer.
.RE
.TP
\fB\-\-thin-results\fR
.RS
Thin Results provides only minimal amount of information in OVAL/ARF results. The option --without-syschar is automatically enabled when you use Thin Results.
.RE
.TP
\fB\-\-without-syschar\fR
.RS
Don't provide system characteristics in OVAL/ARF result files.
.RE
.TP
\fB\-\-report FILE\fR
.RS
Write HTML report into FILE.
.RE
.TP
\fB\-\-oval-results\fR
.RS
Generate OVAL Result file for each OVAL session used for evaluation. File with name '\fIoriginal-oval-definitions-filename\fR.result.xml' will be generated for each referenced OVAL file in current working directory. To change the directory where OVAL files are generated change the CWD using the `cd` command.
.RE
.TP
\fB\-\-check-engine-results\fR
.RS
After evaluation is finished, each loaded check engine plugin is asked to export its results. The export itself is plugin specific, please refer to documentation of the plugin for more details.
.RE
.TP
\fB\-\-export-variables\fR
.RS
Generate OVAL Variables documents which contain external variables' values that were provided to the OVAL checking engine during evaluation. The filename format is '\fIoriginal-oval-definitions-filename\fR-\fIsession-index\fR.variables-\fIvariables-index\fR.xml'.
.RE
.TP
\fB\-\-datastream-id ID\fR
.RS
Uses a data stream with that particular ID from the given data stream collection. If not given the first data stream is used. Only applies if you give source data stream in place of an XCCDF file.
.RE
.TP
\fB\-\-xccdf-id ID\fR
.RS
Takes component ref with given ID from checklists. This allows one to select a particular XCCDF component even in cases where there are multiple XCCDFs in a single data stream. If none is given, the first component from the checklists element is used.
.RE
.TP
\fB\-\-benchmark-id ID\fR
.RS
Selects a component ref from any data stream that references a component with XCCDF Benchmark such that its @id attribute matches given string exactly. Please note that this is not the recommended way of selecting a component-ref. You are advised to use --xccdf-id AND/OR --datastream-id for more precision. --benchmark-id is only used when both --xccdf-id and --datastream-id are not present on the command line!
.RE
.TP
\fB\-\-skip-validation\fR
.RS
Do not validate input/output files.
.RE
.TP
\fB\-\-skip-signature-validation\fR
.RS
Do not validate digital signatures in digitally signed SCAP source data streams.
.RE
.TP
\fB\-\-enforce-signature\fR
.RS
Process only digitally signed SCAP source data streams. Data streams without a signature would be rejected if this switch is used.
.RE
.TP
\fB\-\-fetch-remote-resources\fR
.RS
Allow download of remote OVAL content referenced from XCCDF by check-content-ref/@href.
.RE
.TP
\fB\-\-local-files DIRECTORY\fR
.RS
Instead of downloading remote data stream components from the network, use data stream components stored locally as files in the given directory. In place of the remote data stream component OpenSCAP will attempt to use a file whose file name is equal to @name attribute of the uri element within the catalog element within the component-ref element in the data stream if such file exists.
This option works only with SCAP 1.3 source data streams. It can't be used with SCAP 1.2 source data streams.
.RE
.TP
\fB\-\-remediate\fR
.RS
Execute XCCDF remediation in the process of XCCDF evaluation. This option automatically executes content of XCCDF fix elements for failed rules, and thus this shall be avoided unless for trusted content. Use of this option is always at your own risk.
.RE
.TP
\fB\-\-show-rule-details\fR
.RS
Show rule description and rationale when reporting results.
.RE
.RE
.TP
.B remediate\fR [\fIoptions\fR] INPUT_FILE [\fIoval-definitions-files\fR]
.RS
This module provides post-scan remediation. It assumes that the INPUT_FILE is result of `oscap xccdf eval` operation. The input file must contain TestResult element. This module executes XCCDF fix elements for failed rule-result contained in the given TestResult. Use of this option is always at your own risk and it shall be avoided unless for trusted content.
.TP
\fB\-\-result-id ID
.RS
ID of the XCCDF TestResult element which shall be remedied. If this option is missing the last TestResult (in top-down processing) will be remedied.
.RE
.TP
\fB\-\-skip-validation\fR
.RS
Do not validate input/output files.
.RE
.TP
\fB\-\-fetch-remote-resources\fR
.RS
Allow download of remote OVAL content referenced from XCCDF by check-content-ref/@href.
.RE
.TP
\fB\-\-local-files DIRECTORY\fR
.RS
Instead of downloading remote data stream components from the network, use data stream components stored locally as files in the given directory. In place of the remote data stream component OpenSCAP will attempt to use a file whose file name is equal to @name attribute of the uri element within the catalog element within the component-ref element in the data stream if such file exists.
This option works only with SCAP 1.3 source data streams. It can't be used with SCAP 1.2 source data streams.
.RE
.TP
\fB\-\-cpe CPE_FILE\fR
.RS
Use given CPE dictionary or language (auto-detected) for applicability checks.
.RE
.TP
\fB\-\-results FILE\fR
.RS
Write XCCDF results into FILE.
.RE
.TP
\fB\-\-results-arf FILE\fR
.RS
Writes results to a given FILE in Asset Reporting Format. It is recommended to use this option instead of --results when dealing with data streams.
.RE
.TP
\fB\-\-stig-viewer FILE\fR
.RS
Writes XCCDF results into FILE. The rule result IDs in FILE are modified according to STIG references in evaluated content. The FILE can be simply imported into DISA STIG Viewer. See \fIhttps://public.cyber.mil/stigs/srg-stig-tools/\fR for information about DISA STIG Viewer.
.RE
.TP
\fB\-\-report FILE\fR
.RS
Write HTML report into FILE.
.RE
.TP
\fB\-\-oval-results\fR
.RS
Generate OVAL Result file for each OVAL session used for evaluation. File with name '\fIoriginal-oval-definitions-filename\fR.result.xml' will be generated for each referenced OVAL file.
.RE
.TP
\fB\-\-check-engine-results\fR
.RS
After evaluation is finished, each loaded check engine plugin is asked to export its results. The export itself is plugin specific, please refer to documentation of the plugin for more details.
.RE
.TP
\fB\-\-export-variables\fR
.RS
Generate OVAL Variables documents which contain external variables' values that were provided to the OVAL checking engine during evaluation. The filename format is '\fIoriginal-oval-definitions-filename\fR-\fIsession-index\fR.variables-\fIvariables-index\fR.xml'.
.RE
.TP
\fB\-\-progress\fR
.RS
Switch to sparse output suitable for progress reporting. Format of the output is "$rule_id:$result\\n".
.RE
.TP
\fB\-\-progress-full\fR
.RS
Switch to sparse but a bit more saturated output also suitable for progress reporting. Format of the output is "$rule_id|$rule_title|$result\\n".
.RE
.RE

NEWS
openscap-1.4.3                                                  23-11-2025
- New features
  - Show rule details in output
- Maintenance, bug fix
  - Fix inverted fields in HTML report
  - Add new STIG URL - 1.4
  - Bump OVAL_SUPPORTED to 5.11.3
  - Handle virtual packages in dpkginfo probe
  - Inherit environment for scanning and remediating in `oscap-im` wrapper
  - Fix the way we check rpm library version in CMake
  - Explicitly allow xmlsec1 to load key data from KeyValue
  - Fix double-free in xccdf_policy_remediate.c
  - Fix `unterminated-string-initialization` errors
  - Inherit OpenSCAP environment when executing Bash remediations with `--remediate` option
  - Fix verbose command line option

openscap-1.4.2                                                  06-04-2025
- Maintenance, bug fix
  - Fix thread synchronization bugs
  - Fix textfilecontent54_probe behaviour for negative instance numbers
  - Fix signature obtaining in rpm_info probe

openscap-1.4.1                                                  2025-01-08
- New features
  - Introduce "oscap-im" - script that can be used in Containerfiles to build
    hardened bootable container images to run as Image Mode Operating System
- Maintenance, bug fix
  - Add support for containers with no entrypoint/cmd in "oscap-docker"
  - Stop printing useless component reference information in "oscap info"
  - Fix missing declaration of PATH_MAX on Solaris
  - Fix RPM database path in RPM probes (RHEL-55251, #2151)
  - Fix issues reported by OpenScanHub after 1.4.0 release
  - Fix failing test probes/filehash58/test_probes_filehash58.sh on s390x
    architecture
  - Ensure xlink namespace exists (RHEL-34104)
  - Minor fixes in test suite and CI

openscap-1.4.0                                                  2024-08-13
- New features
  - Introduce ability to generate Kickstarts for unattended OS installation using the `oscap xccdf generate fix --fix-type kickstart` command
  - Add ability to process multi-profile JSON tailorings by the `autotailor` tool
- Removed features
  - Removed `cve`, `cvss`, `cvrf` modules
  - Removed `ds` submodules `sds-compose`, `sds-add`, `sds-split`, `rds-create`, `rds-split`
  - Removed `--template`, `--oval-template` and `--sce-template` options from the `xccdf generate` submodule
  - Remove the `--skip-valid` option (replaced by `--skip-validation`)
- Maintenance, bug fix
  - Advertise path to SSG in remediation scripts
  - Remove the option to build with PCRE
  - Process CPE AL platforms if CPE dictionary isn't part of data stream
  - Disable GConf probe by default (and remove dependencies from docs)
  - Disable MD5 and SHA-1 by default
  - Remove CPE dictionary
  - Fix compiler warnings
  - Update User Manual

openscap-1.3.11                                                 2025-02-10
- New features
  - Introduce "oscap-im" - script that can be used in Containerfiles to build
    hardened bootable container images to run as Image Mode Operating System
- Maintenance, bug fix
  - Fix Python 3.13 compatibility
  - Fix collecting signature in rpm_info probe
  - Fix RPM database path in RPM probes (RHEL-55251, #2151)
  - Ensure xlink namespace exists (RHEL-34104)
  - Stop printing useless component reference information in "oscap info"
  - Minor fixes in test suite and CI

openscap-1.3.10                                                 18-03-2024
- New features
  - Dump all env. variables that affects the behaviour on INFO log level
  - Support Blueprint services customization for masking
  - Fix Blueprint template to be self-contained
  - Add a refine-rule tailoring ability to autotailor
  - Introduce JSON tailoring import option for autotailor
  - Select rules based on reference
  - Skip certain paths from scanning (controlled via env. variable)
  - Introduce a limit of collected items (controlled via env. variable)
- Maintenance, bug fix
  - Fix partition probe for PCRE2
  - Fix NSS crypto backend
  - Wrap Bash snippets in a subshell when generating a fix script
  - Improve references in HTML guides and reports
  - Update html report with OVAL details
  - Rewrite dpkginfo probe without using APT
  - Fix incorrect openscap-cpe-oval result filename
  - Implement xccdf_session_get_rule_results function in XCCDF session API
  - Implement xccdf_session_result_reset function in XCCDF session API

openscap-1.3.9                                                  06-09-2023
- New features
  - OpenSCAP can now use PCRE2 library
- Maintenance, bug fix
  - Fix offline mode (OVAL/sysctl)
  - Fix leak of dpkg cache when dpkginfo_init is called multiple times
  - Fix un-expanded variable in xccdf report output
  - Fix issues when parsing profiles
  - Fix minor problems and resource leaks

openscap-1.3.8                                                  21-06-2023
- New features
  - The boot-time remediation service for systemd's Offline Update mode is now disabled by default
  - Add offline capabilities to the shadow OVAL probe
  - Add offline capabilities to the sysctl OVAL probe
  - Add 'auristorfs' to list of network fileystems
  - Add new experimental linux-bound fwupdsecattr probe for system firmware security attributes (fwupd-based)
- Maintenance, bug fix
  - Use ListUnitFiles D-Bus method to fetch all units in systemd OVAL probe
  - Fix minor resource leaks

openscap-1.3.7                                                  26-01-2023
- Maintenance, bug fix
  - Fix error when processing OVAL filters (rhbz#2126882, rhbz#2126883)
  - Don't emit xmlfilecontent items if XPath doesn't match (rhbz#2138884, rhbz#2139060)
  - Prevent "Failed to check available memory" errors (rhbz#2109485, rhbz#2111040)
  - Make epoch comparison less strict for dpkg
  - Generate graphs when creating Doxygen documentation
  - Fix build on Fedora 37 and Rawhide
  - Fix some compiler warnings
  - Infrastructure and test suite fixes
  - Use more conscious language
  - Fix typos and update documentation


utils/oscap.py
# OVAL simple application for evaluating OVAL definition file.
# The output is system characteristic xml file and oval results xml file.
# Program iterate through tests in every definition and evaluate objects
# of these tests.
#
# Copyright 2010 Red Hat Inc., Durham, North Carolina.
# All Rights Reserved.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful, 
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software 
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Authors:
#      Maros Barabas  <mbarabas@redhat.com>

from openscap_api import *
import sys, getopt, string
import os.path

VERBOSE = 0
LINE_LEN = 80

COLORS = {  '':'', "black":"30", "red":"31", "green":"32", "brown":"33", "blue":"34",
        "purple":"35", "cyan":"36", "lgray":"37", "gray":"1;30", "lred":"1;31",
        "lgreen":"1;32", "yellow":"1;33", "lblue":"1;34", "pink":"1;35",
        "lcyan":"1;36", "white":"1;37" }

XCCDF_RESULT = ["", "green", "red", "lred", "gray", "white", "white", "white", "blue", "yellow"]
OVAL_RESULT = {0:"", 1:"green", 2:"red", 4:"gray", 8:"lred", 16:"white", 32:"white"}

def _color(color, text):
    if sys.stdout.isatty():
        return "\033[%sm%s\033[0m" % (COLORS[color], text)
    else: return text

def _bold(text):
    if sys.stdout.isatty():
        return "\033[1m%s\033[0;0m" % text
    else: return text

def print_save(prefix, indent, string):

    if len(string) < LINE_LEN:
        print prefix, indent, string
    else:
        first = 0
        last_word_len = len(string.split()[-1])
        for n in range(len(string)/LINE_LEN+1):
            last = string[first+LINE_LEN:].find(" ") + first+LINE_LEN
            if last > (len(string)-last_word_len):
                last = len(string)
            s = string[first:last]
            first = last+1
            if len(s) > 0:
                print indent, s

class OSCAP_Action(object):

    def __init__(self):
        self.f_xccdf = None
        self.f_oval = None
        self.f_results = None
        self.url_xccdf = None
        self.url_oval = None
        self.profile = None
        self.cvss_metrics = None
        self.file_version = None
        self.doctype = None


class XCCDF_Handler(object):

    HINT = {"eval" : "Eval: Iterate throught XCCDF Rules and evaluate specified XCCDF Profile upon these rules",
            "validate-xml" : "Validate XCCDF Benchmark file" }

    """
    Init XCCDF Handler with specified parameters
    @param f_oval OVAL Definition file (mandatory)
    @param f_xccdf XCCDF Content file (mandatory)
    @param f_results XCCDF Result file (optional)
    @param profile ID of XCCDF Profile that will be evaluated (optional)
    """
    def __init__(self, args):
 
        self.action = self.__get_opts(args)
        if self.action == None: return

        self.objs = xccdf.init(self.action.f_xccdf)

        self.policy_model = self.objs["policy_model"]
        self.sessions = self.objs["sessions"]
        self.def_models = self.objs["def_models"]
        self.benchmark = self.policy_model.benchmark

        self.__set_policy(self.action.profile)

    def __get_opts(self, arguments):

        action = OSCAP_Action()
        if string.lower(arguments[0]) == "eval":
            setattr(self, "eval", self.__evaluate)
        elif string.lower(arguments[0]) == "validate-xml":
            setattr(self, "eval", self.__validate_xml)
        elif string.lower(arguments[0]) in ("-h", "--help"):
            self.help()
            sys.exit(0)
        else: raise AttributeError("Bad XCCDF operation '%s'" % (arguments[0],))

        try:
            opts, args = getopt.getopt(arguments[1:], "-h", ["help", "result-file=", "profile=", "xccdf-profile=", "file-version="])
        except getopt.GetoptError, err:
            # print help information and exit:
            print str(err) # will print something like "option -a not recognized"
            sys.exit(2)
        for opt, arg in opts:
            if opt in ("-h", "--help"):
                if XCCDF_Handler.HINT.has_key(arguments[0]): self.help(XCCDF_Handler.HINT[arguments[0]])
                else: self.help("No help for option \"xccdf %s\"" % (arguments[0],))
                sys.exit(0)
            elif opt == "--result-file":
                action.f_results = arg
            elif opt in ["--xccdf-profile", "--profile"]:
                action.profile = arg
            elif opt == "--file-version":
                action.file_version = arg
            else:
                assert False, "unhandled option %s" % (opt,)

        if len(args) < 1: raise AttributeError("Need at least XCCDF file (%s parameters given)" %(len(args)))
        action.f_xccdf = args[0]
        action.f_ovals = []
        for arg in args[1:]:
            action.f_ovals.append(arg)
        # TODO Check files

        return action

    def __callback_start(self, msg, plugin):
        result = msg.user2num
        if result == OSCAP.XCCDF_RESULT_NOT_SELECTED: 
            return 0

        print_save( "\nRule ID:", "\r\t\t", _bold(msg.user1str) )
        print_save( "Title:", "\r\t\t", msg.user3str )
        print_save( "Descroption:", "\r\t\t", msg.string )
        print "Result:", "\r\t\t",
        sys.stdout.flush()
        return 0

    def __callback_end(self, msg, plugin):
        result = msg.user2num
        if result == OSCAP.XCCDF_RESULT_NOT_SELECTED: 
            return 0

        print _bold(_color(XCCDF_RESULT[result], xccdf.test_result_type.get_text(msg.user2num)))
        return 0

    def __set_policy(self, policy_id):
        if policy_id == None: 
            self.policy = self.policy_model.policies[0]
            return

        self.policy = None
        policy_list = self.policy_model.policies
        for policy in policy_list:
            if policy.id == policy_id:
                self.policy = policy
        if self.policy == None: raise AttributeError("No policy with '%s' profile id in Policy Model" % (policy_id,))

    def export(self, result=None):
        assert self.action.f_results != None, "Result file not specified"
        if result == None: result = self.result

        files = self.policy.export(result, self.objs, "OSCAP Test Result", self.action.f_results, self.action.f_results)
        return

    def __evaluate(self, policy_id=None):
        if policy_id != None: self.__set_policy(policy_id)
        self.policy_model.register_start_callback(self.__callback_start, None)
        self.policy_model.register_output_callback(self.__callback_end, None)
        self.result = self.policy.evaluate()

        return self.result

    def destroy(self):
        xccdf.destroy(self.objs)

    def __validate_xml(self):
        raise NotImplementedError

    def help(self, msg=None):

        print "Usage: %s [general-options] xccdf command [command-options] OVAL-DEFINITIONS-FILE XCCDF-FILE \n" \
		"(Specify the --help global option for a list of other help options)\n" \
		"\n" \
		"OVAL-DEFINITIONS-FILE is the OVAL XML file specified either by the full path to the xml file " \
		"or an URL from which to download it.\n" \
		"XCCDF-FILE is the XCCDF XML file specified either by the full path to the xml file " \
		"or an URL from which to download it.\n" \
		"\n" \
		"Commands:\n" \
		"   eval\r\t\t\t\t - Perform evaluation driven by XCCDF file and use OVAL as checking engine.\n" \
		"   validate-xml\r\t\t\t\t - validate XCCDF XML content.\n" \
		"\n" \
		"Command options:\n" \
		"   -h --help\r\t\t\t\t - show this help\n" \
		"   --result-file <file>\r\t\t\t\t - Write XCCDF Results into file.\n" \
		"   --profile <name>\r\t\t\t\t - The name of Profile to be evaluated.\n" % (sys.argv[0],)
        if msg != None:
            print "\n%s" % (msg,)


class OVAL_Handler(object):

    HINT = {"eval" : "Eval: Evaluate OVAL Definition file and create results",
            "validate-xml" : "Validate OVAL Definition file" }

    def __init__(self, args):
 
        self.action = self.__get_opts(args)
        if self.action == None: return
        
        self.def_model = oval.definition_model_import(self.action.f_oval)
        assert self.def_model != None and self.def_model.instance != None, "Import definition model failed: %s" % (f_oval,)
        self.sys_model = oval.syschar_model(self.def_model)
        self.pb_sess   = oval.probe_session(self.sys_model)
        self.sess      = oval.agent.new_session(self.def_model, os.path.basename(self.action.f_oval))
        assert self.sess != None and self.sess.instance != None, "Session initialization failed: %s" % (self.action.f_xccdf,)
        self.res_model = oval.agent.results_model(self.sess)

    def __get_opts(self, arguments):

        action = OSCAP_Action()
        if string.lower(arguments[0]) == "eval":
            setattr(self, "eval", self.__evaluate)
        elif string.lower(arguments[0]) == "collect":
            setattr(self, "eval", self.__collect)
        elif string.lower(arguments[0]) == "validate-xml":
            setattr(self, "eval", self.__validate_xml)
        elif string.lower(arguments[0]) in ("-h", "--help"):
            self.help()
            sys.exit(0)
        else: raise AttributeError("Bad OVAL operation '%s'" % (arguments[0],))

        try:
            opts, args = getopt.getopt(arguments[1:], "h", ["help", "result-file=", "file-version=", "--definitions", "--syschar", "--results"])
        except getopt.GetoptError, err:
            # print help information and exit:
            print str(err) # will print something like "option -a not recognized"
            sys.exit(2)
        for opt, arg in opts:
            if opt in ("-h", "--help"):
                if OVAL_Handler.HINT.has_key(arguments[0]): self.help(OVAL_Handler.HINT[arguments[0]])
                else: self.help("No help for option \"oval %s\"" % (arguments[0],))
                sys.exit(0)
            elif opt == "--result-file":
                action.f_results = arg
            elif opt == "--file-version":
                action.file_version = arg
            elif opt == "--definition":
                action.doctype = "definition"
            elif opt == "--syschar":
                action.doctype = "syschar"
            elif opt == "--results":
                action.doctype = "results"
            else:
                assert False, "unhandled option %s" % (opt,)

        if len(args) != 1: raise AttributeError("Need exactly 1 mandatory parameter: OVAL file (%s given)" %(len(args)))
        action.f_oval = args[0]
        # TODO Check files

        return action

    def __callback(self, msg, plugin):
        print "\nID: \r\t\t", _bold(msg.user1str)
        print_save( "Title:", "\r\t\t", msg.user3str )
        print_save( "Description:", "\r\t\t", msg.string )
        print "Result: \r\t\t", _bold(_color(OVAL_RESULT[msg.user2num], oval.result.get_text(msg.user2num)))
        return 0

    def __evaluate(self):
        self.result = oval.agent_eval_system(self.sess, self.__callback, None)

    def __collect(self):
        ret = self.pb_sess.query_sysinfo()
        if ret != 0: return -1
        ret = self.pb_sess.query_objects()
        if ret != 0: return -1

        self.sys_model.export("/dev/stdout")
        return 0

    def __validate_xml(self):
        raise NotImplementedError

    def export(self):
        res_direct = oval.result_directives(self.res_model)
        res_direct.set_reported(OSCAP.OVAL_RESULT_TRUE | OSCAP.OVAL_RESULT_FALSE | OSCAP.OVAL_RESULT_UNKNOWN | OSCAP.OVAL_RESULT_NOT_EVALUATED |
                              OSCAP.OVAL_RESULT_ERROR | OSCAP.OVAL_RESULT_NOT_APPLICABLE, True)
        res_direct.set_content(OSCAP.OVAL_RESULT_FALSE, OSCAP.OVAL_DIRECTIVE_CONTENT_FULL)
        res_direct.set_content(OSCAP.OVAL_RESULT_TRUE, OSCAP.OVAL_DIRECTIVE_CONTENT_FULL)
        self.res_model.export(res_direct, self.action.f_results)

    def help(self, msg=None):
        
        print "Usage: %s [general-options] oval command [command-options] OVAL-DEFINITIONS-FILE \n" \
                "(Specify the --help global option for a list of other help options)\n" \
                "\n" \
                "OVAL-DEFINITIONS-FILE is the OVAL XML file specified either by the full path to the xml file " \
                "or an URL from which to download it.\n" \
                "\n" \
                "Commands:\n" \
                "   collect\r\t\t\t\t - Probe the system and gather system characteristics for objects in OVAL Definition file.\n" \
                "   eval\r\t\t\t\t - Probe the system and evaluate all definitions from OVAL Definition file\n" \
                "   validate-xml\r\t\t\t\t - validate OVAL XML content.\n" \
                "\n" \
                "Command options:\n" \
                "   -h --help\r\t\t\t\t - show this help\n" \
                "   --result-file <file>\r\t\t\t\t - Write OVAL Results into file.\n" % (sys.argv[0],)
        if msg != None:
            print "\n%s" % (msg,)


class CVSS_Handler(object):

    def __init__(self, args):
        print "Sorry, CVSS module is not implemented yet. Please use compiled program utils/oscap instead."
        sys.exit(1)


def help():
    print "Usage: %s [general-options] module operation [operation-options-and-arguments]\n" \
            "\n" \
            "General options:\n" \
            "   -h --help\r\t\t\t\t - show this help\n" \
            "   -q --quiet\r\t\t\t\t - quiet mode\n" \
            "   -V --version\r\t\t\t\t - print info about supported SCAP versions\n" \
            "\n" "Module specific help\n" " %s [oval|xccdf|cvss] --help\n" % (sys.argv[0], sys.argv[0])

def versions():
	print "OSCAP util (oscap) 0.5.12\n","Copyright 2009,2010 Red Hat Inc., Durham, North Carolina.\n"
	print "OVAL Version: \r\t\t%s" % (oval.version)
	print "XCCDF Version: \r\t\t%s" % (xccdf.version)
	print "CVSS Version: \r\t\t%s" % (cvss.version)

def main():
    if len(sys.argv) < 2:
        help()
        sys.exit(1)

    arguments = sys.argv[1:]
    try:
        opts, args = getopt.getopt(arguments, "+qhV", ["quiet", "help", "version"])
    except getopt.GetoptError, err:
        # print help information and exit:
        print str(err) # will print something like "option -a not recognized"
        sys.exit(2)
    for o, a in opts:
        if o in ("-V", "--version"):
            versions()
            sys.exit(0)
            #print version
        elif o in ("-h", "--help"):
            help()
            sys.exit(0)
        elif o in ("-q", "--quiet"):
            VERBOSE = -1
        else:
            assert False, "unhandled option"
    # In args we have all arguments left after opts processing
    if string.lower(args[0]) == "xccdf":
        module = XCCDF_Handler(args[1:])
    elif string.lower(args[0]) == "oval":
        module = OVAL_Handler(args[1:])
    elif string.lower(args[0]) == "cvss":
        module = CVSS_Handler(args[1:])
    else: raise AttributeError("Bad module option '%s'" %(args[0],))

    module.eval()
    if module.action.f_results != None: module.export()
    module.destroy()
    
if __name__ == "__main__":
    main()
